File: Java/Maximizing-Profit/Buy-and-Sell-Stock-I/OnePassPointer.java
class OnePassPointer {
    public int maxProfit(int[] prices) {
        //special case handler
        if (prices.length < 2){
            return 0;
        }
        int buyingPrice = prices[0];
        int maxProfit = 0;
        for (int i = 1; i < prices.length; ++i){
            int profit = prices[i] - buyingPrice;
            if (maxProfit < profit){
                maxProfit = profit;
            }
            else if (prices[i] < buyingPrice){
                buyingPrice = prices[i];
            }
        }
        return maxProfit;
    }
}

File: Python/Maximizing-Profit/Buy-Sell-Stock-I/OnePassPointer.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_profit = 0
        for i in range(1, len(prices)):
            profit = prices[i] - prices[0]
            if max_profit < profit:
                max_profit = profit
            elif prices[i] < prices[0]:
                prices[0] = prices[i]
        return max_profit
